# SKU Creation

# Business Problem:
As per my conversation with Allen Shimon, StrobesnMore needs a merge of their products and product
options so that every possible combination of options has a unique skew. The challenge that their
website currently faces is that their inventory lookup requires visiting multiple links and clicking into
mulƟple dropdowns, making their current list ineffective at looking up product quickly.

`Example:
Product 123 has 2 options with multiple types.`
- Option 1(product feature): Magnetic VS Permanent
- Option 2(color): Blue VS Green

CURRENTLY: Product 123 only has 1 skew (123)
After the complete skew list is created, the following skews will be provided:
- 123MBLUE
- 123PBLUE
- 123MGREEN
- 123PGREEN


Data Understanding
Two csv files were provided to demonstrate the need for this service. There is a csv with a full list of products and another with a list of product options. The goal will be to provide a query merging these two csv files so that each product has a skew associated with every combination of options associated with it. This process will require “digging into” the data and coding a query to pull the list of products and merge all their combination of options. To ensure quality, an understanding of the data options and their products will be required. In the product options list, a few observations have been made and need to be confirmed…

1. It is assumed that each product can only have one unique color.
2. The optionid is the identifier for the option categories for each product. It is assumed that each product can only have one unique option (for example, cannot be both magnetic and permanent) per optionid.
3. The catalogid, NOT the productid in the csv appears to the proper unique identifier for each product.
- `**Edit** SKUs will be created with the productid, not the catalogid, along with the partnumbers.`
4. Client should list if there is a preffered format for SKU creation(color first, capitals, etc)
- `**Edit** Client has issued that he needs the SKUs to be an exact match with the website.`

# Google Colab

This code wad executed in google colab, so the block of code below is the libraries needed to mount the notebook to the users google drive. Depending on the path to your file, the users may be different. 

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


# Libraries Used 

Below are the libraries needed to run the script. 

- The pandas library was used to manipulate the initial spreadsheet. 
- The itertools product was used to compile the permutations (AKA the SKUs) for all the products. 

In [1]:
import pandas as pd
import numpy as np
from itertools import product


pd.set_option('display.max_rows', None)
pd.set_option("display.max_colwidth", 100)

### Functions

Here we have the functions that will obtain the dataframe of SKUs. The process is as follows:

1) get_tuples_for_all_categories, get_tuples_for_all_labels:

These functions get us the partnumbers and featurenames for all the categories associated with each product. The function returns tuples of the partnumber and the sorting number, which in theory is how the partnumbers should be ordered for the SKUs.

2) generate_combinations:

The combination function generates all combinations of the partnumbers, with consideration of order using the 'sorting' column from the dataframe. The functions creates lists of all the possible outcomes and then returns the largest list as some sets of SKUs come out with 1, 2, or 3 options(more options can be added to the function if need be).

3) create_combo_dataframe:

This is the function that puts all these parts together. It iterates through the dataframe and gets all the labels, partnumbers, and combinations and then appends them to a new dataframe composed of the catalogid, productid, the SKU, and the description of the SKU.

4) get_max_unique_partnumbers
This function, get_max_unique_partnumbers, aims to identify the catalog ID (CatalogID) within a given DataFrame (df) that has the maximum number of unique sorting values. It begins by specifying the column names for the catalog ID and the sorting values. Then, it iterates through each unique catalog ID in the DataFrame, filters the DataFrame to obtain data for each catalog ID, and calculates the number of unique sorting values for each catalog ID. During this process, it tracks the catalog ID with the maximum unique sorting values and the corresponding count. Finally, it prints out or returns the catalog ID with the maximum number of unique sorting values along with the count. This function provides a way to identify which catalog ID has the most diverse range of sorting values, which may be useful for various analytical or data exploration purposes.

In [ ]:

def get_tuples_for_all_categories(df):
    # Extract unique category values from the 'cat_w_option' column
    unique_categories = df['cat_w_option'].unique()
    # Initialize an empty list to store tuples of data for each category
    data_by_category = []

    # Iterate through each unique category value
    for category_value in unique_categories:
        # Extract pairs of 'OptionPartNumber' and 'OptionSorting' for the current category
        partnumber_sorting_pairs = list(zip(
            df[df['cat_w_option'] == category_value]['OptionPartNumber'],
            df[df['cat_w_option'] == category_value]['OptionSorting']
        ))
        # Extend the data_by_category list with the extracted pairs
        data_by_category.extend(partnumber_sorting_pairs)

    # Return the list containing tuples of data for all categories
    return data_by_category


def get_tuples_for_all_labels(df):
    # Extract unique category values from the 'cat_w_option' column
    unique_categories = df['cat_w_option'].unique()
    # Initialize an empty list to store tuples of data for each category
    data_by_category = []

    # Iterate through each unique category value
    for category_value in unique_categories:
        # Extract pairs of 'OptionName' and 'OptionSorting' for the current category
        partnumber_sorting_pairs = list(zip(
            df[df['cat_w_option'] == category_value]['OptionName'],
            df[df['cat_w_option'] == category_value]['OptionSorting']
        ))
        # Extend the data_by_category list with the extracted pairs
        data_by_category.extend(partnumber_sorting_pairs)

    # Return the list containing tuples of data for all categories
    return data_by_category



def generate_combinations(input_list):
    # Initialize empty lists to store elements for each unique value
    l1 = []
    l2 = []
    l3 = []
    l4 = []
    l5 = []
    l6 = []
    l7 = []
    l8 = []
    l9 = []
    l10 = []
    l11 = []
    l12 = []

    # Extract values (second elements) from input_list and sort them
    values = [i[1] for i in input_list]
    values = sorted(values)
    # Get unique values from the sorted list
    unique_values = list(set(values))

    # Iterate through input_list
    for i in input_list:
        # Distribute elements from input_list into appropriate lists based on their second element
        if i[1] == unique_values[0]:
            l1.append(i[0])
        elif i[1] == unique_values[1]:
            l2.append(i[0])
        elif i[1] == unique_values[2]:
            l3.append(i[0])
        elif i[1] == unique_values[3]:
            l4.append(i[0])
        elif i[1] == unique_values[4]:
            l5.append(i[0])
        elif i[1] == unique_values[5]:
            l6.append(i[0])
        elif i[1] == unique_values[6]:
            l7.append(i[0])
        elif i[1] == unique_values[7]:
            l8.append(i[0])
        elif i[1] == unique_values[8]:
            l9.append(i[0])
        elif i[1] == unique_values[9]:
            l10.append(i[0])
        elif i[1] == unique_values[10]:
            l11.append(i[0])
        elif i[1] == unique_values[11]:
            l12.append(i[0])

    # Generate combinations based on the number of unique values
    if len(unique_values) == 1:
        perm_1 = l1
        return perm_1
    elif len(unique_values) == 2:
        perm_1_2 = list(product(l1, l2))
        return perm_1_2
    elif len(unique_values) == 3:
        perm_1_2_3 = list(product(l1, l2, l3))
        return perm_1_2_3
    elif len(unique_values) == 4:
        perm_1_2_3_4 = list(product(l1, l2, l3, l4))
        return perm_1_2_3_4
    elif len(unique_values) == 5:
        perm_1_2_3_4_5 = list(product(l1, l2, l3, l4, l5))
    elif len(unique_values) == 6:
        perm_1_2_3_4_5_6 = list(product(l1, l2, l3, l4, l5, l6))
        return perm_1_2_3_4_5_6
    elif len(unique_values) == 7:
        perm_1_2_3_4_5_6_7 = list(product(l1, l2, l3, l4, l5, l6, l7))
        return perm_1_2_3_4_5_6_7
    elif len(unique_values) == 8:
        perm_1_2_3_4_5_6_7_8 = list(product(l1, l2, l3, l4, l5, l6, l7, l8))
        return perm_1_2_3_4_5_6_7_8
    elif len(unique_values) == 9:
        perm_1_2_3_4_5_6_7_8_9 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9))
        return perm_1_2_3_4_5_6_7_8_9
    elif len(unique_values) == 10:
        perm_1_2_3_4_5_6_7_8_9_10 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9, l10))
        return perm_1_2_3_4_5_6_7_8_9_10
    elif len(unique_values) == 11:
        perm_1_2_3_4_5_6_7_8_9_10_11 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9, l10, l11))
        return perm_1_2_3_4_5_6_7_8_9_10_11
    elif len(unique_values) == 12:
        perm_1_2_3_4_5_6_7_8_9_10_11_12 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9, l10, l11, l12))
        return perm_1_2_3_4_5_6_7_8_9_10_11_12



def create_combo_dataframes(df):
    # DataFrame df with columns 'catalogid', 'productid', 'partnumber', 'sorting', 'labels'
    unique_values = df['CatalogID'].unique()

    # Initialize an empty DataFrame to store the results
    combined_df = pd.DataFrame()

    # Iterate through unique values in the specified column
    for value in unique_values:
        # Create a subset DataFrame based on the column value
        subset_df = df[df['CatalogID'] == value]

        # Apply your functions to the subset DataFrame
        result_PN = get_tuples_for_all_categories(subset_df)
        result_labels = get_tuples_for_all_labels(subset_df)
        combination_PN = generate_combinations(result_PN)
        combination_labels = generate_combinations(result_labels)

        # Concatenate the results for each value
        combo_PN = pd.DataFrame(combination_PN).sum(axis=1)
        combo_labels = pd.DataFrame(combination_labels).apply(lambda row: '_'.join(row), axis=1)

        # Add 'productid' column
        name_column = pd.Series(subset_df['Name'].unique(), name='Name')
        product_id_column = pd.Series(subset_df['SKU'].unique(), name='ProductID')
        catalog_id_column = pd.Series(subset_df['CatalogID'].unique(), name='CatalogID')

        # Fill NaN values in combo_PN
        combo_PN = combo_PN.fillna('')

        # Convert float values to strings
        combo_PN = combo_PN.astype(str)
        product_id_column = product_id_column.astype(str)

        # Concatenate all columns
        combo_df = pd.concat([catalog_id_column, product_id_column, combo_PN,name_column, combo_labels], axis=1)

        # Concatenate to the overall DataFrame
        combined_df = pd.concat([combined_df, combo_df])

    # Set columns outside the loop
    combined_df.columns = ['CatalogID', 'ProductID', 'SKUs','Name', 'Labels']
    combined_df = combined_df.fillna(method='ffill')
    combined_df['SKU'] = combined_df['ProductID'] + combined_df['SKUs']
    final_df = combined_df[['CatalogID','ProductID', 'SKU','Name', 'Labels']]

    return final_df

def get_max_unique_partnumbers(df):
    # Replace 'catalogid_column_name' with the actual column name containing catalogid
    # Replace 'sorting_column_name' with the actual column name containing sorting values

    catalogid_column_name = 'CatalogID'
    sorting_column_name = 'OptionSorting'

    # Create variables to store the catalogid with max unique values and the max count
    max_catalogid = None
    max_unique_count = 0

    # Iterate through unique catalogids in the DataFrame
    for catalogid in df[catalogid_column_name].unique():
        # Filter DataFrame for the current catalogid
        catalogid_data = df[df[catalogid_column_name] == catalogid]

        # Count the number of unique sorting values
        unique_sorting_count = len(catalogid_data[sorting_column_name].unique())

        # Check if the current catalogid has more unique values than the current max
        if unique_sorting_count > max_unique_count:
            max_catalogid = catalogid
            max_unique_count = unique_sorting_count

    # Print or use the max_catalogid and max_unique_count as needed
    print(f"The catalogid with the maximum number of unique sorting values is {max_catalogid} with {max_unique_count} unique values.")





# Reading DataFrame


This block of code performs several data preprocessing steps on an Excel dataset containing options information. 

1. Reads the Excel file located at the specified path into a Pandas DataFrame. 
2. Converts certain columns like 'OptionSetID', 'CatalogID', 'OptionPartNumber', and 'OptionName' to string type to ensure consistent data type handling. The values in the 'OptionSorting' column are incremented by 1. 
3. A new column 'cat_w_option' is created by concatenating 'CatalogID' and 'OptionSetID'. Any 'nan' values in the 'OptionPartNumber' column are replaced with an empty string. 
4. A subset DataFrame named 'df_test' is created for testing purposes, containing rows from index 17 to 60. 
5. ain DataFrame is filtered to exclude rows where the 'Hidden' column has a value of 1, and the first few rows of the resulting DataFrame are displayed. Overall, this block of code prepares the dataset for further analysis by ensuring data consistency, creating new columns, and filtering out irrelevant rows.

In [ ]:
# Define the file path
path = "/content/drive/My Drive/Colab Notebooks/Grades/All_Options.xlsx"

# Read the Excel file into a DataFrame
df_options_ = pd.read_excel(path)

# Convert certain columns to string type
df_options_['OptionSetID'] = df_options_['OptionSetID'].astype(str)
df_options_['CatalogID'] = df_options_['CatalogID'].astype(str)
df_options_['OptionPartNumber'] = df_options_['OptionPartNumber'].astype(str)
df_options_['OptionName'] = df_options_['OptionName'].astype(str)

# Increment values in the 'OptionSorting' column by 1
df_options_['OptionSorting'] = df_options_['OptionSorting'] + 1

# Create a new column 'cat_w_option' by concatenating 'CatalogID' and 'OptionSetID'
df_options_['cat_w_option'] = df_options_['CatalogID'] + '_' + df_options_['OptionSetID']

# Replace 'nan' values in 'OptionPartNumber' with empty string
df_options_['OptionPartNumber'] = df_options_['OptionPartNumber'].apply(lambda x: '' if x=='nan' else x)

# Create a subset DataFrame for testing purposes
df_test = df_options_.iloc[17:61]

# Filter the main DataFrame to include only rows where 'Hidden' column is 0
df_options_ = df_options_[df_options_['Hidden'] == 0]

# Display the first few rows of the DataFrame
df_options_.head()


Unnamed: 0 CatalogID      SKU                                  Name  \
0           4      3474  11.1002  Sho Me Universal On/Off/Flash Switch   
1           5      3474  11.1002  Sho Me Universal On/Off/Flash Switch   
2           6      3475   11.120  Sho Me Low Profile LED Mini Lightbar   
3           7      3475   11.120  Sho Me Low Profile LED Mini Lightbar   
4           8      3475   11.120  Sho Me Low Profile LED Mini Lightbar   

   NotForSale   Hide                 OptionName OptionPartNumber OptionSetID  \
0       False  False         Positive Momentary              .PM        4114   
1       False  False  Negative/Ground Momentary             .GSM        4114   
2       False  False                   Magnetic            0.008        2457   
3       False  False                  Permanent                0        2457   
4       False  False                       Blue              -BB        2458   

   OptionSorting  Hidden cat_w_option  
0              2       0    3474_4114  
1              2       0    3474_4114  
2              1       0    3475_2457  
3              1       0    3475_2457  
4              2       0    3475_2458

In [ ]:
path = "/content/drive/My Drive/Colab Notebooks/Grades/allproducts.xlsx"
df_products = pd.read_excel(path)



df_products['OptionSetID'] = df_products['OptionSetID'].astype(str)
df_products['CatalogID'] = df_products['CatalogID'].astype(str)
df_products['OptionPartNumber'] = df_products['OptionPartNumber'].astype(str)
df_products['OptionName'] = df_products['OptionName'].astype(str)
df_products['OptionSorting'] = df_products['OptionSorting'] + 1
df_products['cat_w_option'] =  df_products['CatalogID'] + '_' + df_products['OptionSetID']
df_products['OptionPartNumber'] = df_products['OptionPartNumber'].apply(lambda x: '' if x=='nan' else x)
# df_test = df_products.iloc[17:61]
df_products = df_products[df_products['Hidden'] == 0]


df_products

Unnamed: 0 CatalogID      SKU                                    Name  \
0           0      3457     RB6T  Whelen Dual Reflector Rota Beam Beacon   
1           1      3457     RB6T  Whelen Dual Reflector Rota Beam Beacon   
4           4      3474  11.1002    Sho Me Universal On/Off/Flash Switch   
5           5      3474  11.1002    Sho Me Universal On/Off/Flash Switch   
6           6      3475   11.120    Sho Me Low Profile LED Mini Lightbar   

                  OptionName OptionPartNumber OptionSetID  OptionSorting  \
0                       Blue               BP        2451              1   
1                      Clear               CP        2451              1   
4         Positive Momentary              .PM        4114              2   
5  Negative/Ground Momentary             .GSM        4114              2   
6                   Magnetic            0.008        2457              1   

   Hidden cat_w_option  
0       0    3457_2451  
1       0    3457_2451  
4       0    3474_4114  
5       0    3474_4114  
6       0    3475_2457

# Checking for NotForSale and NotOnWebsite Products.

During the scraping of the product data, it was discovered that some of the products from the API are listed as not for sale or not on the website. We chose not to include these products in the final SKU list, but kept the list for inventory purposes. 

## Filtering by NotForSale

In [ ]:
NotForSale = df_options_[df_options_['NotForSale'] == True]

## Checking 'Hide' Values

If a product is listed as 'Hide' = True, it is off the website. If listed as False, then it is on the website. 

In [ ]:
df_options_['Hide'].unique()

array([False])

## Checking 'NotForSale' Values

If a product is listed as 'NotForSale' = True, it is off the website. If listed as False, then it is on the website. 

In [ ]:
df_options_['NotForSale'].unique()

array([False,  True])

## All catalogIDs of products not for sale, after filtering. 

In [ ]:
NotForSale['CatalogID'].unique()

array(['5862', '5888', '6743'], dtype=object)

In [ ]:
NotForSale

Unnamed: 0 CatalogID        SKU  \
1816         466      5862      PB400   
1817         467      5862      PB400   
1818         468      5862      PB400   
1819         469      5862      PB400   
1820         470      5862      PB400   
1822         472      5862      PB400   
1823         473      5862      PB400   
1824         474      5862      PB400   
1825         475      5862      PB400   
1826         476      5862      PB400   
1827         477      5862      PB400   
1828         478      5862      PB400   
1829         479      5862      PB400   
1830         480      5862      PB400   
1831         481      5862      PB400   
1832         482      5862      PB400   
1833         483      5862      PB400   
1834         484      5862      PB400   
1835         485      5862      PB400   
1843         493      5888      PB450   
1844         494      5888      PB450   
1845         495      5888      PB450   
1846         496      5888      PB450   
1847         497      5888      PB450   
1848         498      5888      PB450   
1849         499      5888      PB450   
1850         500      5888      PB450   
1851         501      5888      PB450   
1852         502      5888      PB450   
1853         503      5888      PB450   
1854         504      5888      PB450   
1855         505      5888      PB450   
1856         506      5888      PB450   
1857         507      5888      PB450   
1858         508      5888      PB450   
1859         509      5888      PB450   
1860         510      5888      PB450   
1861         511      5888      PB450   
1862         512      5888      PB450   
1863         513      5888      PB450   
1864         514      5888      PB450   
1865         515      5888      PB450   
1866         516      5888      PB450   
1867         517      5888      PB450   
1868         518      5888      PB450   
1869         519      5888      PB450   
1870         520      5888      PB450   
1871         521      5888      PB450   
1872         522      5888      PB450   
1873         523      5888      PB450   
1874         524      5888      PB450   
1875         525      5888      PB450   
1878         528      5888      PB450   
1879         529      5888      PB450   
3352         143      6743  TCRWXTRIO   
3353         144      6743  TCRWXTRIO   

                                                         Name  NotForSale  \
1816                                       Setina Push Bumper        True   
1817                                       Setina Push Bumper        True   
1818                                       Setina Push Bumper        True   
1819                                       Setina Push Bumper        True   
1820                                       Setina Push Bumper        True   
1822                                       Setina Push Bumper        True   
1823                                       Setina Push Bumper        True   
1824                                       Setina Push Bumper        True   
1825                                       Setina Push Bumper        True   
1826                                       Setina Push Bumper        True   
1827                                       Setina Push Bumper        True   
1828                                       Setina Push Bumper        True   
1829                                       Setina Push Bumper        True   
1830                                       Setina Push Bumper        True   
1831                                       Setina Push Bumper        True   
1832                                       Setina Push Bumper        True   
1833                                       Setina Push Bumper        True   
1834                                       Setina Push Bumper        True   
1835                                       Setina Push Bumper        True   
1843                             Setina Push Bumper with LEDs        True   
1844                             Setina Push Bumper

# Removing Items Not for Sale

In [ ]:
df_options_ = df_options_[df_options_['NotForSale'] == False]

# Grabbing Columns

### Checking length of dataframe after filtering by products we intend to get SKUs for

In [ ]:
len(df_options_)

3391

In [ ]:
df_options_.columns

Index(['Unnamed: 0', 'CatalogID', 'SKU', 'Name', 'NotForSale', 'Hide',
       'OptionName', 'OptionPartNumber', 'OptionSetID', 'OptionSorting',
       'Hidden', 'cat_w_option'],
      dtype='object')

## Removing the 'Unnamed: 0' Column

In [ ]:
df_options_ = df_options_[['CatalogID','SKU','OptionName','Name','OptionPartNumber','OptionSetID','OptionSorting','cat_w_option']]

In [ ]:
df_options_.head()

CatalogID      SKU                 OptionName  \
0      3474  11.1002         Positive Momentary   
1      3474  11.1002  Negative/Ground Momentary   
2      3475   11.120                   Magnetic   
3      3475   11.120                  Permanent   
4      3475   11.120                       Blue   

                                   Name OptionPartNumber OptionSetID  \
0  Sho Me Universal On/Off/Flash Switch              .PM        4114   
1  Sho Me Universal On/Off/Flash Switch             .GSM        4114   
2  Sho Me Low Profile LED Mini Lightbar            0.008        2457   
3  Sho Me Low Profile LED Mini Lightbar                0        2457   
4  Sho Me Low Profile LED Mini Lightbar              -BB        2458   

   OptionSorting cat_w_option  
0              2    3474_4114  
1              2    3474_4114  
2              1    3475_2457  
3              1    3475_2457  
4              2    3475_2458

# Checking Datatypes

Here we are making sure the dataframe has the correct datatypes for the functions to run.

In [ ]:
df_options_.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3391 entries, 0 to 3766
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   CatalogID         3391 non-null   object
 1   SKU               3391 non-null   object
 2   OptionName        3391 non-null   object
 3   Name              3391 non-null   object
 4   OptionPartNumber  3391 non-null   object
 5   OptionSetID       3391 non-null   object
 6   OptionSorting     3391 non-null   int64 
 7   cat_w_option      3391 non-null   object
dtypes: int64(1), object(7)
memory usage: 238.4+ KB


# Checking that the max product number of options is up to 8. 

Client has agreed that the cutoff for number of options to be founs for SKUs to be 8. This function checks for the max number of options for the products.

In [ ]:
get_max_unique_partnumbers(df_options_)

The catalogid with the maximum number of unique sorting values is 3939 with 8 unique values.


## Confirming the catalogid provided by the function to check for 8 options

In [ ]:
df_options_[df_options_['CatalogID'] == '3939']

CatalogID SKU OptionName                               Name  \
432      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
433      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
434      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
435      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
436      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
437      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
438      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
439      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
440      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
441      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
442      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
443      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
444      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
445      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
446      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
447      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
448      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
449      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
450      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
451      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
452      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
453      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
454      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
455      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
456      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
457      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
458      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
459      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
460      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
461      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
462      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
463      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   

    OptionPartNumber OptionSetID  OptionSorting cat_w_option  
432                B        2987              2    3939_2987  
433                C        2987              2    3939_2987  
434                R        2987              2    3939_2987  
435                A        2987              2    3939_2987  
436                B        2988              3    3939_2988  
437                C        2988              3    3939_2988  
438                R        2988              3    3939_2988  
439                A        2988              3    3939_2988  
440                B        2989              4    3939_2989  
441                C        2989              4    3939_2989  
442                R        2989              4    3939_2989  
443                A        2989              4    3939_2989  
444                B        2990              5    3939_2990  
445                C        2990              5    3939_2990  
446                R        2990              5    3939_2990  
447                A        2990              5    3939_2990  
448                B        2991              6    3939_2991  
449                C        2991              6    3939_2991  
450                R        2991              6    3939_2991  
451                A        2991              6    3939_2991  
452                B        2992              7    3939_2992  
453                C        2992              7    3939_2992  
454                R        2992              7    3939_2992  
455                A        2992              7    3939_2992  
456                B        2993              8    3939_2993  
457                C        2993              8    3939_2993  
458                R        2993              8    3939_2993  
459                A        

## Checking for unique SKUs - ensuring no errors to avoid mismatch with website

In [ ]:
df_options_[df_options_['OptionPartNumber'] =='']['SKU'].unique()

array(['M4B6', 'SA315U', 'SA350M', 'SC-550', 'SLWIC3', 'SLPMM',
       'IW0BAAAA', 'C-AS-840-11', 'C-EB-', 'HOWLER', 'F4N', 'ES100C',
       'ULB24', 'C-1800', 'C-1810', 'C-2410', 'R316', 'R416', 'ENFLB',
       'C-4014', 'MCRNS', 'LINSV2', 'GALAXYELITE', 'F4M', 'C-4015',
       'C-VS-0412-CAPR-1', 'SA350MH', 'CCSRN', 'LF35', 'FS', 'TCRWXSOLO',
       'TCRWXDUO', 'GMP-F150-2015', 'OELS', 'GALAXYELITE-PURPLE', 'PCP3',
       'C-', 'FS-8816', 'FS-0416', 'FS-0616', 'FS-0816', 'C6T', 'C6BTT',
       'C7T', 'C7BTT', 'C9T', 'C9BTT', 'ENRLB', 'TRIOFST-', 'EMPLB',
       'S-3017', 'C-AP-0325-1', 'C-AP-0625-1', 'C-AP-0645-1',
       'POLICE-ALGT', 'C6BU', 'C7BU', 'PLOWPKG4', 'PF200', 'PCH1', 'PCH2',
       'PHANTOM', 'E3', 'E6', 'E62', 'E64', 'E68', 'E66', 'PF200R',
       'PF200H', 'PF200S17', 'SHADOW', 'CY', 'SNM-PROMOTSHIRT',
       'Q-0420\t', 'Q', '50LENSL', 'C399', 'ARGES', 'CHOWLER', 'ION',
       'WION', 'XON', 'WXON', 'C399R', 'RLNT48', 'INTGPROMO', 'VALRPROMO',
       'ILSPROMO-SIF', 

# Test Run

Here we are checking the create_combo_dataframes function to check that it works on the sample dataframe, taken from the master list. 

In [ ]:
df_test = df_options_.iloc[:106]
df_test.tail()

CatalogID SKU              OptionName                         Name  \
104      3495  90                     Red  Whelen 900 Series Super LED   
105      3495  90     Blue with Blue Lens  Whelen 900 Series Super LED   
106      3495  90       Red With Red Lens  Whelen 900 Series Super LED   
107      3495  90  Amber with Amber Lens   Whelen 900 Series Super LED   
108      3495  90  Green with Green Lens   Whelen 900 Series Super LED   

    OptionPartNumber OptionSetID  OptionSorting cat_w_option  
104           RR5FCR        2479              1    3495_2479  
105           BB5FBR        2479              1    3495_2479  
106           RR5FRR        2479              1    3495_2479  
107           AA5FAR        2479              1    3495_2479  
108           GG5FGR        2479              1    3495_2479

In [ ]:
create_combo_dataframes(df_test)


CatalogID ProductID             SKU  \
0       3474   11.1002      11.1002.PM   
1       3474   11.1002     11.1002.GSM   
0       3475    11.120  11.1200.008-BB   
1       3475    11.120  11.1200.008-RB   
2       3475    11.120  11.1200.008-BC   
3       3475    11.120  11.1200.008-RC   
4       3475    11.120  11.1200.008-AC   
5       3475    11.120  11.1200.008-GG   
6       3475    11.120  11.1200.008-RR   
7       3475    11.120  11.1200.008-AA   
8       3475    11.120  11.1200.008-RA   
9       3475    11.120      11.1200-BB   
10      3475    11.120      11.1200-RB   
11      3475    11.120      11.1200-BC   
12      3475    11.120      11.1200-RC   
13      3475    11.120      11.1200-AC   
14      3475    11.120      11.1200-GG   
15      3475    11.120      11.1200-RR   
16      3475    11.120      11.1200-AA   
17      3475    11.120      11.1200-RA   
0       3479         2        2EB00ZCR   
1       3479         2        2EC00ZCR   
2       3479         2        2ER00ZCR   
3       3479         2        2EA00ZCR   
4       3479         2        2EJ00ZCR   
5       3479         2        2EA00ZAR   
6       3479         2        2EB00ZBR   
7       3479         2        2ER00ZRR   
8       3479         2        2FB00ZCR   
9       3479         2        2FC00ZCR   
10      3479         2        2FR00ZCR   
11      3479         2        2FA00ZCR   
12      3479         2        2FJ00ZCR   
13      3479         2        2FA00ZAR   
14      3479         2        2FB00ZBR   
15      3479         2        2FR00ZRR   
0       3483        68  68-1183725-1SB   
1       3483        68  68-1183725-2SB   
2       3483        68  68-1183725-3SB   
3       3483        68  68-1183725-5SB   
4       3483        68  68-1183726-1SB   
5       3483        68  68-1183726-2SB   
6       3483        68  68-1183726-3SB   
7       3483        68  68-1183726-5SB   
8       3483        68  68-3183725-1SB   
9       3483        68  68-3183725-2SB   
10      3483        68  68-3183725-3SB   
11      3483        68  68-3183725-5SB   
12      3483        68  68-3183726-1SB   
13      3483        68  68-3183726-2SB   
14      3483        68  68-3183726-3SB   
15      3483        68  68-3183726-5SB   
16      3483        68  68-1183725-1SB   
17      3483        68  68-1183725-2SB   
18      3483        68  68-1183725-3SB   
19      3483        68  68-1183725-5SB   
20      3483        68  68-1183726-1SB   
21      3483        68  68-1183726-2SB   
22      3483        68  68-1183726-3SB   
23      3483        68  68-1183726-5SB   
0       3485        50        50AA2ZCR   
1       3485        50        50CR2ZCR   
2       3485        50        50CC2ZCR   
3       3485        50        50RR2ZCR   
4       3485        50        50BB2ZCR   
5       3485        50        50RB2ZCR   
0       3486        50        50BB3ZCR   
1       3486        50        50CR3ZCR   
2       3486        50        50CC3ZCR   
3       3486        50        50RR3ZCR   
4       3486        50        50AA3ZCR   
5       3486        50        50RB3ZCR   
6       3486        50        50AC3ZCR   
7       3486        50        50RA3ZCR   
0       3487        68  68-1963237-50A   
1       3487        68  68-1963237-10A   
2       3487        68  68-1963237-20A   
3       3487        68  68-1963237-30A   
4       3487        68  68-1963583-50A   
5       3487        68  68-1963583-10A   
6       3487        68  68-1963583-20A   
7       3487        68  68-1963583-30A   
8       3487        68  68-1963583-50A   
9       3487        68  68-1963583-10A   
10      3487        68  68-1963583-20A   
11      3487        68  68-1963583-30A   
12      3487        68  68-1963583-50A   
13      3487        68  68-1963583-10A   
14      3487        68  68-1963583-20A   
15      3487        68  68-1963583-30A   
0       3489        60        60C02FCR   
1       3489        60        60A02FCR   
2       3489        60        60BR6FCR   
3       3489        60        60AC6FCR   
4       3489 

## Initial Test Run - Success

After checking through the test dataframe we see that the SKUs created were a match with the website. This is also after checks done in the API as well as the store website. 

## Run on entire SKU list

In [5]:
Master_SKU_list = create_combo_dataframes(df_options_)

## Checking number of rows after run

In [ ]:
len(Master_SKU_list)

279997

Length of dataframe is 279,997 SKUs. Client has approved this length.

## Manual Checks

In the dataframe we checked for the string '--' to see if there were any missing partnumbers that may have caused some SKUs to come out of order. These 5 products below were the ones that were found. The dataframe was adjusted and the product SKUs now match the website. 

In [ ]:
Master_SKU_list[Master_SKU_list['SKU'].str.contains('--')]
#check F4N
#check LegacyDUO
#check LegacySOLO
#check LIBERTYIISOLO
#check SOLOFST-


CatalogID    ProductID                       SKU  \
3        6463     SOLOFST-  SOLOFST-ISFW355--CONTACT   
9        6463     SOLOFST-  SOLOFST-ISFW35Z--CONTACT   
15       6463     SOLOFST-  SOLOFST-ISFW425--CONTACT   
21       6463     SOLOFST-  SOLOFST-ISFW426--CONTACT   
27       6463     SOLOFST-  SOLOFST-ISFW42X--CONTACT   
33       6463     SOLOFST-  SOLOFST-ISFW42Z--CONTACT   
39       6463     SOLOFST-  SOLOFST-ISFW434--CONTACT   
45       6463     SOLOFST-  SOLOFST-ISFW435--CONTACT   
51       6463     SOLOFST-  SOLOFST-ISFW438--CONTACT   
57       6463     SOLOFST-  SOLOFST-ISFW43Z--CONTACT   
63       6463     SOLOFST-  SOLOFST-ISFW444--CONTACT   
69       6463     SOLOFST-  SOLOFST-ISFW445--CONTACT   
75       6463     SOLOFST-  SOLOFST-ISFW448--CONTACT   
81       6463     SOLOFST-  SOLOFST-ISFW44Z--CONTACT   
87       6463     SOLOFST-  SOLOFST-ISFW455--CONTACT   
93       6463     SOLOFST-  SOLOFST-ISFW456--CONTACT   
99       6463     SOLOFST-  SOLOFST-ISFW45X--CONTACT   
105      6463     SOLOFST-  SOLOFST-ISFW45Z--CONTACT   
111      6463     SOLOFST-  SOLOFST-ISFW474--CONTACT   
117      6463     SOLOFST-  SOLOFST-ISFW475--CONTACT   
123      6463     SOLOFST-  SOLOFST-ISFW478--CONTACT   
129      6463     SOLOFST-  SOLOFST-ISFW47Z--CONTACT   
135      6463     SOLOFST-  SOLOFST-ISFW485--CONTACT   
141      6463     SOLOFST-  SOLOFST-ISFW486--CONTACT   
147      6463     SOLOFST-  SOLOFST-ISFW48X--CONTACT   
153      6463     SOLOFST-  SOLOFST-ISFW48Z--CONTACT   
159      6463     SOLOFST-  SOLOFST-ISFW495--CONTACT   
165      6463     SOLOFST-  SOLOFST-ISFW496--CONTACT   
171      6463     SOLOFST-  SOLOFST-ISFW49Z--CONTACT   
177      6463     SOLOFST-  SOLOFST-ISFW49X--CONTACT   
183      6463     SOLOFST-  SOLOFST-ISFW354--CONTACT   
189      6463     SOLOFST-  SOLOFST-ISFW358--CONTACT   
195      6463     SOLOFST-  SOLOFST-ISFW504--CONTACT   
201      6463     SOLOFST-  SOLOFST-ISFW505--CONTACT   
207      6463     SOLOFST-  SOLOFST-ISFW508--CONTACT   
213      6463     SOLOFST-  SOLOFST-ISFW50Z--CONTACT   
219      6463     SOLOFST-  SOLOFST-BSFW355--CONTACT   
225      6463     SOLOFST-  SOLOFST-BSFW35Z--CONTACT   
231      6463     SOLOFST-  SOLOFST-BSFW425--CONTACT   
237      6463     SOLOFST-  SOLOFST-BSFW426--CONTACT   
243      6463     SOLOFST-  SOLOFST-BSFW42X--CONTACT   
249      6463     SOLOFST-  SOLOFST-BSFW42Z--CONTACT   
255      6463     SOLOFST-  SOLOFST-BSFW434--CONTACT   
261      6463     SOLOFST-  SOLOFST-BSFW435--CONTACT   
267      6463     SOLOFST-  SOLOFST-BSFW438--CONTACT   
273      6463     SOLOFST-  SOLOFST-BSFW43Z--CONTACT   
279      6463     SOLOFST-  SOLOFST-BSFW444--CONTACT   
285      6463     SOLOFST-  SOLOFST-BSFW445--CONTACT   
291      6463     SOLOFST-  SOLOFST-BSFW448--CONTACT   
297      6463     SOLOFST-  SOLOFST-BSFW44Z--CONTACT   
303      6463     SOLOFST-  SOLOFST-BSFW455--CONTACT   
309      6463     SOLOFST-  SOLOFST-BSFW456--CONTACT   
315      6463     SOLOFST-  SOLOFST-BSFW45X--CONTACT   
321      6463     SOLOFST-  SOLOFST-BSFW45Z--CONTACT   
327      6463     SOLOFST-  SOLOFST-BSFW474--CONTACT   
333      6463     SOLOFST-  SOLOFST-BSFW475--CONTACT   
339      6463     SOLOFST-  SOLOFST-BSFW478--CONTACT   
345      6463     SOLOFST-  SOLOFST-BSFW47Z--CONTACT   
351      6463     SOLOFST-  SOLOFST-BSFW485--CONTACT   
357      6463     SOLOFST-  SOLOFST-BSFW486--CONTACT   
363      6463     SOLOFST-  SOLOFST-BSFW48X--CONTACT   
369      6463     SOLOFST-  SOLOFST-BSFW48Z--CONTACT   
375      6463     SOLOFST-  SOLOFST-BSFW495--CONTACT   
381      6463     SOLOFST-  SOLOFST-BSFW496--CONTACT   
387      6463     SOLOFST-  SOLOFST-BSFW49Z--CONTACT   
393      6463     SOLOFST-  SOLOFST-BSFW49X--CONTACT   
399      6463     SOLOFST-  SOLOFST-BSFW354--CONTACT   
405      6463     SOLOFST-  SOLOFST-BSFW358--CONTACT   
411      6463     SOLOFST-  SOLOFST-BSFW504--CONTACT   
417      6463     SOLOFST-  SOLOFST-BSFW505--CONTACT   
423      6463     SOLOFST-  

# Checks on the following products:
- #check F4N **done**
- #check LegacyDUO **done**
- #check LegacySOLO **done**
- #check LIBERTYIISOLO **done**
- #check SOLOFST-	**done**

In [ ]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'F4N']

CatalogID ProductID                SKU  \
0        4136       F4N   F4N2-MKEZ95-VLED   
1        4136       F4N   F4N2-MKEZ95-VLED   
2        4136       F4N   F4N2-MKEZ95-VLED   
3        4136       F4N   F4N2-MKEZ95-QLED   
4        4136       F4N   F4N2-MKEZ95-QLED   
5        4136       F4N   F4N2-MKEZ95-QLED   
6        4136       F4N   F4N2-MKEZ96-VLED   
7        4136       F4N   F4N2-MKEZ96-VLED   
8        4136       F4N   F4N2-MKEZ96-VLED   
9        4136       F4N   F4N2-MKEZ96-QLED   
10       4136       F4N   F4N2-MKEZ96-QLED   
11       4136       F4N   F4N2-MKEZ96-QLED   
12       4136       F4N  F4N2-MKEZ101-VLED   
13       4136       F4N  F4N2-MKEZ101-VLED   
14       4136       F4N  F4N2-MKEZ101-VLED   
15       4136       F4N  F4N2-MKEZ101-QLED   
16       4136       F4N  F4N2-MKEZ101-QLED   
17       4136       F4N  F4N2-MKEZ101-QLED   
18       4136       F4N   F4N2-MKEZ45-VLED   
19       4136       F4N   F4N2-MKEZ45-VLED   
20       4136       F4N   F4N2-MKEZ45-VLED   
21       4136       F4N   F4N2-MKEZ45-QLED   
22       4136       F4N   F4N2-MKEZ45-QLED   
23       4136       F4N   F4N2-MKEZ45-QLED   
24       4136       F4N   F4N2-MKEZ91-VLED   
25       4136       F4N   F4N2-MKEZ91-VLED   
26       4136       F4N   F4N2-MKEZ91-VLED   
27       4136       F4N   F4N2-MKEZ91-QLED   
28       4136       F4N   F4N2-MKEZ91-QLED   
29       4136       F4N   F4N2-MKEZ91-QLED   
30       4136       F4N   F4N2-MKEZ93-VLED   
31       4136       F4N   F4N2-MKEZ93-VLED   
32       4136       F4N   F4N2-MKEZ93-VLED   
33       4136       F4N   F4N2-MKEZ93-QLED   
34       4136       F4N   F4N2-MKEZ93-QLED   
35       4136       F4N   F4N2-MKEZ93-QLED   
36       4136       F4N  F4N2-MKAJ101-VLED   
37       4136       F4N  F4N2-MKAJ101-VLED   
38       4136       F4N  F4N2-MKAJ101-VLED   
39       4136       F4N  F4N2-MKAJ101-QLED   
40       4136       F4N  F4N2-MKAJ101-QLED   
41       4136       F4N  F4N2-MKAJ101-QLED   
42       4136       F4N   F4N2-MKEZ69-VLED   
43       4136       F4N   F4N2-MKEZ69-VLED   
44       4136       F4N   F4N2-MKEZ69-VLED   
45       4136       F4N   F4N2-MKEZ69-QLED   
46       4136       F4N   F4N2-MKEZ69-QLED   
47       4136       F4N   F4N2-MKEZ69-QLED   
48       4136       F4N   F4N2-MKEZ82-VLED   
49       4136       F4N   F4N2-MKEZ82-VLED   
50       4136       F4N   F4N2-MKEZ82-VLED   
51       4136       F4N   F4N2-MKEZ82-QLED   
52       4136       F4N   F4N2-MKEZ82-QLED   
53       4136       F4N   F4N2-MKEZ82-QLED   
54       4136       F4N   F4N2-MKEZ85-VLED   
55       4136       F4N   F4N2-MKEZ85-VLED   
56       4136       F4N   F4N2-MKEZ85-VLED   
57       4136       F4N   F4N2-MKEZ85-QLED   
58       4136       F4N   F4N2-MKEZ85-QLED   
59       4136       F4N   F4N2-MKEZ85-QLED   
60       4136       F4N   F4N2-MKEZ62-VLED   
61       4136       F4N   F4N2-MKEZ62-VLED   
62       4136       F4N   F4N2-MKEZ62-VLED   
63       4136       F4N   F4N2-MKEZ62-QLED   
64       4136       F4N   F4N2-MKEZ62-QLED   
65       4136       F4N   F4N2-MKEZ62-QLED   
66       4136       F4N   F4N2-MKAJ62-VLED   
67       4136       F4N   F4N2-MKAJ62-VLED   
68       4136       F4N   F4N2-MKAJ62-VLED   
69       4136       F4N   F4N2-MKAJ62-QLED   
70       4136       F4N   F4N2-MKAJ62-QLED   
71       4136       F4N   F4N2-MKAJ62-QLED   
72       4136       F4N   F4N2-MKEZ86-VLED   
73       4136       F4N   F4N2-MKEZ86-VLED   
74       4136       F4N   F4N2-MKEZ86-VLED   
75       4136       F4N   F4N2-MKEZ86-QLED   
76       4136       F4N   F4N2-MKEZ86-QLED   
77       4136       F4N   F4N2-MKEZ86-QLED   
78       4136       F4N  F4N2-MKEZ102-VLED   
79       4136       F4N  F4N2-MKEZ102-VLED   
80       4136       F4N  F4N2-MKEZ102-VLED   
81       4136       F4N  F4N2-MKEZ102-QLED   
82       4136       F4N  F4N2-MKEZ102-QLED   
83       4136       F4N  F4N2-MKEZ102-QLED   
84       4136       F4N  F4N2-MKEZ107-VLED   
85       4136       F4N  F4N2-MKEZ107-VLED   
86

# F4N - checked

In [ ]:
df_options_[df_options_['CatalogID'] == '4311' ]

CatalogID        SKU  \
829      4311  LEGACYDUO   
830      4311  LEGACYDUO   
831      4311  LEGACYDUO   
832      4311  LEGACYDUO   
833      4311  LEGACYDUO   
834      4311  LEGACYDUO   
835      4311  LEGACYDUO   
836      4311  LEGACYDUO   
837      4311  LEGACYDUO   
838      4311  LEGACYDUO   
839      4311  LEGACYDUO   
840      4311  LEGACYDUO   
841      4311  LEGACYDUO   
842      4311  LEGACYDUO   
843      4311  LEGACYDUO   
844      4311  LEGACYDUO   
845      4311  LEGACYDUO   
846      4311  LEGACYDUO   
847      4311  LEGACYDUO   
848      4311  LEGACYDUO   
849      4311  LEGACYDUO   
850      4311  LEGACYDUO   
851      4311  LEGACYDUO   
852      4311  LEGACYDUO   
853      4311  LEGACYDUO   
854      4311  LEGACYDUO   
855      4311  LEGACYDUO   
856      4311  LEGACYDUO   
857      4311  LEGACYDUO   
858      4311  LEGACYDUO   
859      4311  LEGACYDUO   
860      4311  LEGACYDUO   
861      4311  LEGACYDUO   
862      4311  LEGACYDUO   
863      4311  LEGACYDUO   
864      4311  LEGACYDUO   
865      4311  LEGACYDUO   
866      4311  LEGACYDUO   
867      4311  LEGACYDUO   
868      4311  LEGACYDUO   
869      4311  LEGACYDUO   
870      4311  LEGACYDUO   
871      4311  LEGACYDUO   
872      4311  LEGACYDUO   
873      4311  LEGACYDUO   
874      4311  LEGACYDUO   
875      4311  LEGACYDUO   
876      4311  LEGACYDUO   
877      4311  LEGACYDUO   
878      4311  LEGACYDUO   
879      4311  LEGACYDUO   
880      4311  LEGACYDUO   
881      4311  LEGACYDUO   
882      4311  LEGACYDUO   
883      4311  LEGACYDUO   
884      4311  LEGACYDUO   
885      4311  LEGACYDUO   

                                                                                              OptionName  \
829                                                                                            48 Inches   
830                                                                                            54 Inches   
831                                                                 Red = Red/White Front Red/Amber Rear   
832                                                             Blue = Blue/White Front. Blue/Amber Rear   
833                      Red/Blue = Half Red Half Blue w/ White Front - Half Red Half Blue w/ Amber Rear   
834                                                           Amber = Amber/White Front Amber/White Rear   
835                                                           WeCan Control *Require WeCan Control Point   
836                                                    WeCanX Control Point *Requires CORE Control Point   
837                   Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”   
838                                                     Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"   
839                                                                 Silverado 1500 / 2021-2023 / 54"-56"   
840                                                Silverado Pickup / 1999-2006 (2007 Classic) / 54"-56"   
841                                            Silverado 1500 / GMC 1500 Pickup / 2014-2018 / 54"-56"+B6   
842                                                                          Tahoe / 2015-2020 / 54"-56"   
843                               Tahoe / 2021-2023 / 54"-56" (Not for Use with Factory Roof Rack Rails)   
844                                                                        Charger / 2005-2010 / 48"-50"   
845                                                                        Charger / 2011-2023 / 48"-50"   
846                                                                        Durango / 2011-2023 / 48"-50"   
847                                                                        Durango / 2004-2010 / 48"-50"   
848                                              Durango, Adjustable Leveling Foot / 2004-2010 / 48"-50"   
849                  Ram 1500 / 2012-2018 / Ram 1500 Classic / 2019-2021 / Ram 2500/3500 / 2019/ 54"-56"   
850

In [ ]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'LEGACYDUO']

CatalogID  ProductID                            SKU  \
0        4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ95-RT   
1        4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ95-BT   
2        4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ95-JT   
3        4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ95-FFFF   
4        4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ96-RT   
5        4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ96-BT   
6        4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ96-JT   
7        4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ96-FFFF   
8        4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ101-RT   
9        4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ101-BT   
10       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ101-JT   
11       4311  LEGACYDUO  LEGACYDUO-GB8SP3-MKEZ101-FFFF   
12       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ45-RT   
13       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ45-BT   
14       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ45-JT   
15       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ45-FFFF   
16       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ91-RT   
17       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ91-BT   
18       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ91-JT   
19       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ91-FFFF   
20       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ93-RT   
21       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ93-BT   
22       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ93-JT   
23       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ93-FFFF   
24       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKAJ101-RT   
25       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKAJ101-BT   
26       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKAJ101-JT   
27       4311  LEGACYDUO  LEGACYDUO-GB8SP3-MKAJ101-FFFF   
28       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ69-RT   
29       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ69-BT   
30       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ69-JT   
31       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ69-FFFF   
32       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ82-RT   
33       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ82-BT   
34       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ82-JT   
35       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ82-FFFF   
36       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ85-RT   
37       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ85-BT   
38       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ85-JT   
39       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ85-FFFF   
40       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ62-RT   
41       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ62-BT   
42       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ62-JT   
43       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ62-FFFF   
44       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKAJ62-RT   
45       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKAJ62-BT   
46       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKAJ62-JT   
47       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKAJ62-FFFF   
48       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ86-RT   
49       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ86-BT   
50       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ86-JT   
51       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ86-FFFF   
52       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ102-RT   
53       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ102-BT   
54       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ102-JT   
55       4311  LEGACYDUO  LEGACYDUO-GB8SP3-MKEZ102-FFFF   
56       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ107-RT   
57       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ107-BT   
58       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ107-JT   
59       4311  LEGACYDUO  LEGACYDUO-GB8SP3-MKEZ107-FFFF   
60       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ89-RT   
61       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ89-BT   
62       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ89-JT   
63       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ89-FFFF   
64       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ38-RT   
65       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ38-BT   
66       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ3

# LegacyDUO - checked


In [ ]:
df_options_[df_options_['CatalogID'] == '4413' ]

CatalogID         SKU  \
965       4413  LEGACYSOLO   
966       4413  LEGACYSOLO   
967       4413  LEGACYSOLO   
968       4413  LEGACYSOLO   
969       4413  LEGACYSOLO   
970       4413  LEGACYSOLO   
971       4413  LEGACYSOLO   
972       4413  LEGACYSOLO   
973       4413  LEGACYSOLO   
974       4413  LEGACYSOLO   
975       4413  LEGACYSOLO   
976       4413  LEGACYSOLO   
977       4413  LEGACYSOLO   
978       4413  LEGACYSOLO   
979       4413  LEGACYSOLO   
980       4413  LEGACYSOLO   
981       4413  LEGACYSOLO   
982       4413  LEGACYSOLO   
983       4413  LEGACYSOLO   
984       4413  LEGACYSOLO   
985       4413  LEGACYSOLO   
986       4413  LEGACYSOLO   
987       4413  LEGACYSOLO   
988       4413  LEGACYSOLO   
989       4413  LEGACYSOLO   
990       4413  LEGACYSOLO   
991       4413  LEGACYSOLO   
992       4413  LEGACYSOLO   
993       4413  LEGACYSOLO   
994       4413  LEGACYSOLO   
995       4413  LEGACYSOLO   
996       4413  LEGACYSOLO   
997       4413  LEGACYSOLO   
998       4413  LEGACYSOLO   
999       4413  LEGACYSOLO   
1000      4413  LEGACYSOLO   
1001      4413  LEGACYSOLO   
1002      4413  LEGACYSOLO   
1003      4413  LEGACYSOLO   
1004      4413  LEGACYSOLO   
1005      4413  LEGACYSOLO   
1006      4413  LEGACYSOLO   
1007      4413  LEGACYSOLO   
1008      4413  LEGACYSOLO   
1009      4413  LEGACYSOLO   
1010      4413  LEGACYSOLO   
1011      4413  LEGACYSOLO   
1012      4413  LEGACYSOLO   
1013      4413  LEGACYSOLO   
1014      4413  LEGACYSOLO   
1015      4413  LEGACYSOLO   
1016      4413  LEGACYSOLO   
1017      4413  LEGACYSOLO   
1018      4413  LEGACYSOLO   
1019      4413  LEGACYSOLO   
1020      4413  LEGACYSOLO   

                                                                                               OptionName  \
965                                                                                  Half Red / Half Blue   
966                                                                                                 Amber   
967                                                                                                  Blue   
968                                                                                                   Red   
969                                                                                             48 Inches   
970                                                                                             54 Inches   
971                                                            WeCan Control *Require WeCan Control Point   
972                                                         WeCanX Conttrol *Requires CORE Control Point    
973                    Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”   
974                                                      Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"   
975                                                                  Silverado 1500 / 2021-2023 / 54"-56"   
976                                                 Silverado Pickup / 1999-2006 (2007 Classic) / 54"-56"   
977                                             Silverado 1500 / GMC 1500 Pickup / 2014-2018 / 54"-56"+B6   
978                                                                           Tahoe / 2015-2020 / 54"-56"   
979                                Tahoe / 2021-2023 / 54"-56" (Not for Use with Factory Roof Rack Rails)   
980                                                                         Charger / 2005-2010 / 48"-50"   
981                                                                         Charger / 2011-2023 / 48"-50"   
982                                                                         Durango / 2011-2023 / 48"-50"   
983                                                                         Durango / 2004-2010 / 48"-50"   
984                                               Durango, Adjustable Leveling Foot / 2004-2010 / 48"-50"   
985  

In [ ]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'LEGACYSOLO']

CatalogID   ProductID                          SKU  \
0        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-J   
1        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-A   
2        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-B   
3        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-R   
4        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-J   
5        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-A   
6        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-B   
7        4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-R   
8        4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-J   
9        4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-A   
10       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-B   
11       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-R   
12       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-J   
13       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-A   
14       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-B   
15       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-R   
16       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-J   
17       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-A   
18       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-B   
19       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-R   
20       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-J   
21       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-A   
22       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-B   
23       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-R   
24       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-J   
25       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-A   
26       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-B   
27       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-R   
28       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-J   
29       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-A   
30       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-B   
31       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-R   
32       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-J   
33       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-A   
34       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-B   
35       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-R   
36       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-J   
37       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-A   
38       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-B   
39       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-R   
40       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-J   
41       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-A   
42       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-B   
43       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-R   
44       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-J   
45       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-A   
46       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-B   
47       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-R   
48       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-J   
49       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-A   
50       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-B   
51       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-R   
52       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-J   
53       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-A   
54       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-B   
55       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-R   
56       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-J   
57       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-A   
58       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-B   
59       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-R   
60       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-J   
61       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-A   
62       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-B   
63       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-R   
64       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-J   
65       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-A   
66       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-B   
67       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-R   
68

# Legacy Solo - Checked

In [ ]:
df_options_[df_options_['CatalogID'] == '6174']

CatalogID            SKU  \
2438      6174  LIBERTYIISOLO   
2439      6174  LIBERTYIISOLO   
2440      6174  LIBERTYIISOLO   
2441      6174  LIBERTYIISOLO   
2442      6174  LIBERTYIISOLO   
2443      6174  LIBERTYIISOLO   
2444      6174  LIBERTYIISOLO   
2445      6174  LIBERTYIISOLO   
2446      6174  LIBERTYIISOLO   
2447      6174  LIBERTYIISOLO   
2448      6174  LIBERTYIISOLO   
2449      6174  LIBERTYIISOLO   
2450      6174  LIBERTYIISOLO   
2451      6174  LIBERTYIISOLO   
2452      6174  LIBERTYIISOLO   
2453      6174  LIBERTYIISOLO   
2454      6174  LIBERTYIISOLO   
2455      6174  LIBERTYIISOLO   
2456      6174  LIBERTYIISOLO   
2457      6174  LIBERTYIISOLO   
2458      6174  LIBERTYIISOLO   
2459      6174  LIBERTYIISOLO   
2460      6174  LIBERTYIISOLO   
2461      6174  LIBERTYIISOLO   
2462      6174  LIBERTYIISOLO   
2463      6174  LIBERTYIISOLO   
2464      6174  LIBERTYIISOLO   
2465      6174  LIBERTYIISOLO   
2466      6174  LIBERTYIISOLO   
2467      6174  LIBERTYIISOLO   
2468      6174  LIBERTYIISOLO   
2469      6174  LIBERTYIISOLO   
2470      6174  LIBERTYIISOLO   
2471      6174  LIBERTYIISOLO   
2472      6174  LIBERTYIISOLO   
2473      6174  LIBERTYIISOLO   
2474      6174  LIBERTYIISOLO   
2475      6174  LIBERTYIISOLO   
2476      6174  LIBERTYIISOLO   
2477      6174  LIBERTYIISOLO   
2478      6174  LIBERTYIISOLO   
2479      6174  LIBERTYIISOLO   
2480      6174  LIBERTYIISOLO   
2481      6174  LIBERTYIISOLO   
2482      6174  LIBERTYIISOLO   
2483      6174  LIBERTYIISOLO   
2484      6174  LIBERTYIISOLO   
2485      6174  LIBERTYIISOLO   
2486      6174  LIBERTYIISOLO   
2487      6174  LIBERTYIISOLO   
2488      6174  LIBERTYIISOLO   
2489      6174  LIBERTYIISOLO   
2490      6174  LIBERTYIISOLO   
2491      6174  LIBERTYIISOLO   
2492      6174  LIBERTYIISOLO   
2493      6174  LIBERTYIISOLO   
2494      6174  LIBERTYIISOLO   

                                                                                               OptionName  \
2438                                                                                                   48   
2439                                                                                                   54   
2440                                                                                                Amber   
2441                                                                                                 Blue   
2442                                                                                                  Red   
2443                                                                                             Red/Blue   
2444                                               Contact Me for Colors (This may come with an upcharge)   
2445                                                           WeCan Control *Require WeCan Control Point   
2446                                                    WeCanX Control Point *Requires CORE Control Point   
2447                   Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”   
2448                                                     Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"   
2449                                                                 Silverado 1500 / 2021-2023 / 54"-56"   
2450                                                Silverado Pickup / 1999-2006 (2007 Classic) / 54"-56"   
2451                                            Silverado 1500 / GMC 1500 Pickup / 2014-2018 / 54"-56"+B6   
2452                                                                          Tahoe / 2015-2020 / 54"-56"   
2453                               Tahoe / 2021-2023 / 54"-56" (Not for Use with Factory Roof Rack Rails)   
2454                                                                        Charger / 2005-2010 / 48"-50"   
2455                                                                        Charger / 2011-2023 / 48"-50"   
2456               

In [ ]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'LIBERTYIISOLO']

CatalogID      ProductID                             SKU  \
0        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-AAAA   
1        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-BBBB   
2        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-RRRR   
3        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-BRBR   
4        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-Call   
5        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-AAAA   
6        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-BBBB   
7        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-RRRR   
8        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-BRBR   
9        6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-Call   
10       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-AAAA   
11       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-BBBB   
12       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-RRRR   
13       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-BRBR   
14       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-Call   
15       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-AAAA   
16       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-BBBB   
17       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-RRRR   
18       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-BRBR   
19       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-Call   
20       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ91-AAAA   
21       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ91-BBBB   
22       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ91-RRRR   
23       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ91-BRBR   
24       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ91-Call   
25       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ93-AAAA   
26       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ93-BBBB   
27       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ93-RRRR   
28       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ93-BRBR   
29       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ93-Call   
30       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKAJ101-AAAA   
31       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKAJ101-BBBB   
32       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKAJ101-RRRR   
33       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKAJ101-BRBR   
34       6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKAJ101-Call   
35       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ69-AAAA   
36       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ69-BBBB   
37       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ69-RRRR   
38       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ69-BRBR   
39       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ69-Call   
40       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ82-AAAA   
41       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ82-BBBB   
42       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ82-RRRR   
43       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ82-BRBR   
44       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ82-Call   
45       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ85-AAAA   
46       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ85-BBBB   
47       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ85-RRRR   
48       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ85-BRBR   
49       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ85-Call   
50       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ62-AAAA   
51       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ62-BBBB   
52       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ62-RRRR   
53       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ62-BRBR   
54       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ62-Call   
55       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKAJ62-AAAA   
56       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKAJ62-BBBB   
57       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKAJ62-RRRR   
58       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKAJ62-BRBR   
59       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKAJ62-Call   
60       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ86-AAAA   
61       6174  LIBERTYIISOLO   LIBER

# LIBERTYIISOLO - Checked

Note - SKU was behaving very strangely during SKU compilation.

In [ ]:
Master_SKU_list.tail(1000)

In [ ]:
Master_SKU_list[Master_SKU_list['SKU'].str.endswith('.0')]['CatalogID'].unique()

array(['3500', '3794', '3980', '4087', '4089', '4278', '4471', '5352',
       '5411', '5441', '5543', '5613', '5614', '5745', '5890', '5899',
       '5947', '5957', '6031', '6032', '6037', '6038', '6041', '6042',
       '6043', '6292', '6297', '6337', '6366', '6388', '6406', '6408',
       '6667', '6826', '6950', '6979'], dtype=object)

In [ ]:
Master_SKU_list[Master_SKU_list['SKU'].str.endswith('.0')]

CatalogID    ProductID                SKU  \
0        3500        ALPHA           ALPHA1.0   
1        3500        ALPHA           ALPHA2.0   
2        3500        ALPHA           ALPHA3.0   
3        3500        ALPHA           ALPHA4.0   
4        3500        ALPHA           ALPHA5.0   
0        3794        PA300     PA300-690000.0   
1        3794        PA300     PA300-690010.0   
0        3980          WCC             WCC9.0   
1        3980          WCC            WCC92.0   
0        4087          905            90540.0   
1        4087          905             9053.0   
0        4089       NSP-11         NSP-1170.0   
1        4089       NSP-11         NSP-1172.0   
2        4089       NSP-11         NSP-1174.0   
0        4278        CCSRN           CCSRN3.0   
1        4278        CCSRN          CCSRN35.0   
2        4278        CCSRN          CCSRN36.0   
0        4471           PC              PC2.0   
1        4471           PC              PC3.0   
2        4471           PC              PC1.0   
0        5352       IONBKT          IONBKT7.0   
1        5352       IONBKT          IONBKT9.0   
0        5411         TAC8           TAC815.0   
1        5411         TAC8           TAC830.0   
2        5411         TAC8           TAC850.0   
0        5441        C-FP-           C-FP-5.0   
1        5441        C-FP-           C-FP-1.0   
2        5441        C-FP-          C-FP-15.0   
3        5441        C-FP-           C-FP-2.0   
4        5441        C-FP-          C-FP-25.0   
5        5441        C-FP-           C-FP-3.0   
6        5441        C-FP-          C-FP-35.0   
7        5441        C-FP-           C-FP-4.0   
8        5441        C-FP-           C-FP-6.0   
9        5441        C-FP-           C-FP-5.0   
10       5441        C-FP-           C-FP-8.0   
11       5441        C-FP-          C-FP-10.0   
0        5543        TADSW           TADSW1.0   
1        5543        TADSW           TADSW2.0   
0        5613         OEWD           OEWD45.0   
0        5614         OEWT           OEWT45.0   
0        5745           C-           C-5017.0   
1        5745           C-           C-5017.0   
2        5745           C-           C-5017.0   
3        5745           C-           C-4017.0   
4        5745           C-           C-4017.0   
5        5745           C-           C-4017.0   
6        5745           C-           C-4017.0   
7        5745           C-           C-4017.0   
8        5745           C-           C-4017.0   
0        5890        ENRLB         ENRLB-48.0   
1        5890        ENRLB         ENRLB-48.0   
2        5890        ENRLB         ENRLB-54.0   
3        5890        ENRLB         ENRLB-54.0   
0        5899        EMPLB         EMPLB-44.0   
1        5899        EMPLB         EMPLB-44.0   
2        5899        EMPLB         EMPLB-44.0   
3        5899        EMPLB         EMPLB-44.0   
4        5899        EMPLB         EMPLB-48.0   
5        5899        EMPLB         EMPLB-48.0   
6        5899        EMPLB         EMPLB-48.0   
7        5899        EMPLB         EMPLB-48.0   
8        5899        EMPLB         EMPLB-53.0   
9        5899        EMPLB         EMPLB-53.0   
10       5899        EMPLB         EMPLB-53.0   
11       5899        EMPLB         EMPLB-53.0   
0        5947          UT-          UT-2001.0   
1        5947          UT-          UT-2003.0   
2        5947          UT-          UT-2004.0   
3        5947          UT-          UT-2005.0   
4        5947          UT-          UT-2006.0   
5        5947          UT-          UT-2007.0   
6        5947          UT-          UT-2008.0   
7        5947          UT-          UT-2009.0   
8        5947          UT-          UT-2010.0   
0        5957       LSVBKT         LSVBKT13.0   
1        5957       LSVBKT         LSVBKT35.0   
2        5957       LSVBKT         LSVBKT44.0   
3        5957       LSVBKT         LSVBKT42.0   
4        5957       LSVBKT         LSVBKT47.0   
5        5957       LSVBKT         

## Removing the '.0'

After checks, it was noticed that some of the SKUs came in with an extra '.0' at the end. This happened with SKUs ending in a numerical value. These were removed to complete the match. 

In [ ]:
def remove_decimal(value):
    if value.endswith('.0'):
        return value[:-2]  # Remove the last two characters (.0)
    return value

# Apply the function to the column where you want to remove '.0'
Master_SKU_list['SKU'] = Master_SKU_list['SKU'].apply(remove_decimal)

In [ ]:
Master_SKU_list[Master_SKU_list['SKU'].str.endswith('.0')]['CatalogID'].unique()

array([], dtype=object)

All '.0' parts of the SKUs were successfully removed. 

# Here, Master_SKU_List is exported to a CSV

In [ ]:
Master_SKU_list.to_csv('Master_SKU_list.csv')

In [ ]:
len(Master_SKU_list)

279997

In [ ]:
Master_SKU_list.columns

Index(['CatalogID', 'ProductID', 'SKU', 'Name', 'Labels'], dtype='object')

In [ ]:
Master_SKU_list.isnull().sum()

CatalogID    0
ProductID    0
SKU          0
Name         0
Labels       0
dtype: int64

# Conclusion

In this project, we have created a Master SKU List of all products from the Strobe's N More Website. 

This list was compiled and pushed to a CSV for the client, along with a written report and a list of products off the website. 

## Future Work


Client was informed via the written report that the API contains some issues of ordering that made manual checks necessary. Client was advised to make the changes in the API to overwrite for future Inventory issues. 